src/shared/tpm2-util.c

```diff
@@ -663,60 +818,25 @@ const TPM2B_PUBLIC *tpm2_get_primary_template(Tpm2SRKTemplateFlags flags) {
  */
 static int tpm2_get_srk(
                 Tpm2Context *c,
+                const Tpm2Handle *session,
                 TPMI_ALG_PUBLIC *ret_alg,
-                Tpm2Handle *ret_primary) {
+                Tpm2Handle **ret_handle) {
 
-        TPMI_YES_NO more_data;
-        ESYS_TR primary_tr = ESYS_TR_NONE;
-        _cleanup_(Esys_Freep) TPMS_CAPABILITY_DATA *cap_data = NULL;
+        int r;
 
         assert(c);
-        assert(ret_primary);
 
-        TSS2_RC rc = sym_Esys_GetCapability(c->esys_context,
-                        ESYS_TR_NONE,
-                        ESYS_TR_NONE,
-                        ESYS_TR_NONE,
-                        TPM2_CAP_HANDLES,
-                        SRK_HANDLE,
-                        1,
-                        &more_data,
-                        &cap_data);
-        if (rc != TSS2_RC_SUCCESS)
-                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
-                                       "Failed to enumerate handles searching for SRK: %s",
-                                       sym_Tss2_RC_Decode(rc));
-
-        /* Did Not find SRK, indicate this by returning 0 */
-        if (cap_data->data.handles.count == 0 || cap_data->data.handles.handle[0] != SRK_HANDLE) {
-                ret_primary->esys_handle = ESYS_TR_NONE;
-
-                if (ret_alg)
-                        *ret_alg = 0;
-                return 0;
-        }
-
-        log_debug("Found SRK on TPM.");
-
-        /* convert the raw handle to an ESYS_TR */
-        TPM2_HANDLE handle = cap_data->data.handles.handle[0];
-        rc = sym_Esys_TR_FromTPMPublic(c->esys_context,
-                        handle,
-                        ESYS_TR_NONE,
-                        ESYS_TR_NONE,
-                        ESYS_TR_NONE,
-                        &primary_tr);
-        if (rc != TSS2_RC_SUCCESS)
-                return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
-                                        "Failed to convert ray handle to ESYS_TR for SRK: %s",
-                                        sym_Tss2_RC_Decode(rc));
+        _cleanup_(tpm2_handle_freep) Tpm2Handle *handle = NULL;
+        r = tpm2_esys_handle_from_tpm_handle(c, session, SRK_HANDLE, &handle);
+        if (r <= 0) /* error or handle not found */
+                return r;
```

Comment:
similar to above, all the `*ret_xyz` should be initialized here in case of r == 0

Reply:
updated